### PR TITLE
Json to recordset

### PR DESCRIPTION
**Pull request by aljungberg**
Switches from `json_populate_recordset` to `json_to_recordset`, which we will need for data representations. See #2523 for details.

The impossible panic we discussed in #2523 has been eliminated by creating a `TypedField` type which doesn't allow that situation.

### CODE REVIEW

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs:106]: I think there are n ways to create an invalid field description like this, not just an empty string but invalid identifiers and so on. So there's no need to go to extremes with validation, which is why I didn't. At the same time a little validation can save some headache down the road, fail fast style.

Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs:106]: Actually I managed to do it while making this change with a typo in the SQL such that the field type ended up indeed empty! I can also imagine some request with `columns=,b` or whatever.

Anyhow I don't have a strong opinion either way. Happy to make the constructor public if that's preferred.

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs:113]: Agreed. Would be nice if the linter caught such inconsistencies.

File: src/PostgREST/SchemaCache/Table.hs
Comment (aljungberg) [reply src/PostgREST/SchemaCache/Table.hs:68]: Great, I like it.

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs]: Fixed in the next commit.

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs:111]: Hmm yes. I wanted it near `Field`. We will end up with some analogs such as `Filter` vs `TypedFilter` (which leads to `TypedLogicTree` etc too unfortunately). It is true though that like the defining difference is sort of that in `ApiRequest` we have kind of raw user defined data and later we've distilled that into augmented, better validated data. So based on that extinction, certainly putting it in `Plan/` makes a lot of sense because it will be `Plan` that's responsible for that distillation.

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs:106]: Right, isn't that the idea with `TypedField` though, to be better validated and therefore less able to cause runtime errors down the line? The nature of bugs is that they seem impossible, otherwise you wouldn't have them!

Just making the point for the sake of rational debate, again, not married to having extra error checking here.

File: src/PostgREST/Plan/Types.hs
Comment (aljungberg) [reply src/PostgREST/Plan/Types.hs:32]: `resolveField` is used by `resolveTableField` to deconstruct the `Field`. Keeps it cleaner, and I expect to use it when we do need the `jsonPath` in TypedField in the next PR (for selects).

**Comment by aljungberg** [comment]
Regarding the code coverage on `Ord`, what's the best approach there? I could stick a few tests in that do some `compare` exercises, but didn't see anywhere else we do that. Could just drop `Ord` from table, seems to be unused, which would mean we don't need it on `ColumnMap` either. We only need `Hashable` for the maps. Thoughts?

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs]: Has now been moved.

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs]: I made the constructor public while moving everything to `Plan`.

File: src/PostgREST/ApiRequest/Types.hs
Comment (aljungberg) [reply src/PostgREST/ApiRequest/Types.hs]: I dropped the sanity checking from `TypedField`.

**Comment by aljungberg** [comment]
Not entirely sure what this means:

Screenshot 2022-10-28 at 16 09 00

Clearly `insCol` is tested since it's used by the insertion tests?

**Comment by steve-chavez** [comment]
> Switches from json_populate_recordset to json_to_recordset

Would it be possible to fallback to `json_populate_recordset` in case the columns of a relation are not present on the schema cache?

For example, this could happen if a new table is added and the schema cache is not refreshed. We remained operational before for these cases, what would happen now?

**Comment by aljungberg** [comment]
> > Switches from json_populate_recordset to json_to_recordset
> 
> Would it be possible to fallback to `json_populate_recordset` in case the columns of a relation are not present on the schema cache?
> 
> For example, this could happen if a new table is added and the schema cache is not refreshed. We remained operational before for these cases, what would happen now?

The new behaviour is that you get an error when trying to insert data into a table or a column we believe not to exist.

Personally I think that's actually an improvement over the old behaviour. Clear and immediate messaging for a situation that is more likely to be an error than not.

Like how often do you expect to make API calls to tables so fresh off the presses they're not even in the cache? The only time I can imagine is if you're live developing your schema. In that case can't you do `NOTIFY pgrst, 'reload schema'` at the end of your edit transaction as well?

What's the venn diagram between people live editing their schema yet unwilling to `NOTIFY`? I imagine it's pretty small.

- If we make it a stated goal to always allow interaction with unknown tables and columns, data representations will have confusing failure modes (e.g. they work some of the time).
- Even if you don't use data representations, you can have confusing behaviour with regards to relations, embedding and so forth if you are in the habit of changing your schema without reloading the cache. 
- There's more complexity in our codebase having fallbacks.
- To support `json_populate_recordset` as a fallback, the constraints on `TypedField` would have to be relaxed. The record started out relaxed and I made it more strict due to the feedback on [the data rep PR](https://github.com/PostgREST/postgrest/pull/2523). To be clear, I'm happy to do the work, but we might be pulling in two directions at once here.

All that said, happy to change it. I'm really excited about this feature and I think it'll be a great addition to PostgREST, so keen to get it merged. We're going to be using this probably in every single API on our end.

File: src/PostgREST/Error.hs
Comment (steve-chavez) [inline src/PostgREST/Error.hs:188]: Should we include on `details` the list of detected columns? Similarly to how we do it for an error on a relationship?

https://github.com/PostgREST/postgrest/blob/b2935ef7990107642e261b25b6f30a2aed932864/test/spec/Feature/Query/EmbedDisambiguationSpec.hs#L17-L44

**Comment by steve-chavez** [comment]
Cool, I agree that becoming more strict is good overall. 

Not sure if stuffing schema cache info on the `details` key of the error message is the right approach. The schema cache is becoming more relevant with this change and maybe we should make it more visible. 

How about we add a special header on the endpoints to get their schema cache entry(discussed before on https://github.com/PostgREST/postgrest/issues/1421).

```http
GET /projects
Accept: application/vnd.pgrst.schema-cache

{
  "columns": [".."],
  "relationships": [".."]
}
```

---

The above can be done on another PR but I'd consider it a must for a stable release after this feature is merged.

File: src/PostgREST/Error.hs
Comment (aljungberg) [reply src/PostgREST/Error.hs:188]: I like it! Can't see a security problem either, the columns of tables/views exposed through PostgREST are already published in other ways.

**Comment by aljungberg** [comment]
> Not sure if stuffing schema cache info on the `details` key of the error message is the right approach. The schema cache is becoming more relevant with this change and maybe we should make it more visible.
> 
> How about we add a special header on the endpoints to get their schema cache entry(discussed before on #1421).
> 
> ```httpspec
> GET /projects
> Accept: application/vnd.pgrst.schema-cache
> 
> {
>   "columns": [".."],
>   "relationships": [".."]
> }
> ```
> 
> The above can be done on another PR but I'd consider it a must for a stable release after this feature is merged.

Having a per relation info API would be a nice addition. 

As you rightfully point out in the referenced PR, the schema cache is already visible through the root endpoint so perhaps such an API would be more of a nice to have? 

Maybe there's something we can do there to present more information about the available data representations for each column. Haven't fully thought that through yet, but if we add support for CSV and binary data representations down the line it might be nice to see which column custom formatters and/or parsers are available, and which ones will rely on the base type native PostgreSQL casting.

**Comment by steve-chavez** [comment]
> How about making a request to the root endpoint with a different accept header, i.e. the one you mentioned.. and then returning schema cache instead of OpenAPI output?

Yeah, that could be another option. I've branched off the discussion to https://github.com/PostgREST/postgrest/issues/1421#issuecomment-1300683598 to not make this thread longer.

**Comment by aljungberg** [comment]
> Are we throwing the same kind of error messages before? In that case the feature is probably "throwing those errors **without hitting the database**"?

The essence of the error is the same but the error message now has a PostgREST error code and indeed happens without hitting the database.

```
# Before
{"code":"42703","details":null,"hint":null,"message":"column \"helicopter\" of relation \"articles\" does not exist"}
# After
{"code":"PGRST118","details":null,"hint":"If a new column was created in the database with this name, try reloading the schema cache.","message":"Could not find 'helicopter' in the target table"}
```

File: src/PostgREST/Plan.hs
Comment (steve-chavez) [inline src/PostgREST/Plan.hs:381]: This looks like [traverse](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:traverse)

> (a -> f b) -> t a -> f (t b)

You might be able to use it like 

```haskell
typedColumnsOrError = resolveOrError (HM.lookup qi $ dbTables sCache) `traverse` S.toList iColumns
```

---

There is an example on https://github.com/PostgREST/postgrest/blob/78d45b4e32e1129f25c2b6739dbd587ef873822c/src/PostgREST/Plan.hs#L310

File: src/PostgREST/Plan.hs
Comment (steve-chavez) [inline src/PostgREST/Plan.hs:367]: This looks up the table for each column right? Maybe we should cache the table like so
 
```suggestion
    tbl = HM.lookup qi $ dbTables sCache
    typedColumnsOrError = mapStopOnLeft (resolveOrError tbl) $ S.toList iColumns
```

---

Checking the [load tests](https://github.com/PostgREST/postgrest/pull/2542/checks?check_run_id=9363756197)

param | main | head
-- | -- | --
throughput | 317.3776065823794 | 308.3469937995588

We're losing some perf on the PR. Perhaps the above will help.

File: src/PostgREST/Error.hs
Comment (steve-chavez) [reply src/PostgREST/Error.hs:188]: A more useful `hint` for this one. 

When referring to a bad column name, postgres tries to detect a typo:

```
$ curl 'localhost:3000/projects?select=id,nam'
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"projects.name\".","message":"column projects.nam does not exist"}

$ curl 'localhost:3000/projects?select=id,namee'
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"projects.name\".","message":"column projects.namee does not exist"}

$ curl 'localhost:3000/projects?select=id,nae'
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"projects.name\".","message":"column projects.nae does not exist"}

$ curl 'localhost:3000/projects?select=id,xxname'
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"projects.name\".","message":"column projects.xxname does not exist"}

## here it no longer suggest a column
$ curl 'localhost:3000/projects?select=id,yyxxname'
{"code":"42703","details":null,"hint":null,"message":"column projects.yyxxname does not exist"}
```

PostgreSQL does some fuzzy logic for this: [errorMissingColumn](https://github.com/postgres/postgres/blob/5035c93c8a5ac6804da79c67403460348b381924/src/backend/parser/parse_relation.c#L3625-L3696).

Perhaps we can use https://hackage.haskell.org/package/fuzzyfind-3.0.0/docs/Text-FuzzyFind.html for something similar?

File: src/PostgREST/Query/QueryBuilder.hs
Comment (steve-chavez) [inline src/PostgREST/Query/QueryBuilder.hs:93]: This reminds me of https://github.com/PostgREST/postgrest/issues/1567#issuecomment-739478169, not having defaults is still a problem.

Would `json_to_recordset` help or make it harder to do the above suggestion later on?

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs:367]: Great comment, glad to see this level of focus on performance. That said, I think no, it doesn't look up the table more than once, more on that below.

First let me just comment on the performance impact overall of this part of the PR: I think the impact will be limited since the operation just adds a small constant time lookup per field, which should be dominated by the time already spent per field. The stuff postgresql does like parsing the JSON, parsing the SQL, doing the actual work of writing to the table; all of these have complexity linearish to _n_, _n_ being the column count, and each component alone is likely much larger than the time for this op (a hash map lookup in RAM).

Back to the table lookup. I'd expect GHC to only evaluate the `HM.lookup qi $ dbTables sCache` node once while partially applying `resolveOrError`. I'm no Haskell guy so might just be flopping about like a fish out of water here, but take a look at the generated core (`ghc -fforce-recomp -O2 -ddump-simpl`):

```
typedColumnsOrError_a2geX
  = hpc<PostgREST.Plan,179>
    hpc<PostgREST.Plan,178>
    $ (hpc<PostgREST.Plan,175>
       mapStopOnLeft_r2g9q
         (let {
            ds20_d2gYE :: Maybe Table
            [LclId]
            ds20_d2gYE
              = hpc<PostgREST.Plan,173>
                $ (hpc<PostgREST.Plan,170>
                   HM.lookup
                     PostgREST.SchemaCache.Identifiers.$fEqQualifiedIdentifier
                     PostgREST.SchemaCache.Identifiers.$fHashableQualifiedIdentifier
                     (hpc<PostgREST.Plan,169> qi_a2geh))
                  (hpc<PostgREST.Plan,172>
                   dbTables (hpc<PostgREST.Plan,171> sCache1_a2geC)) } in
          hpc<PostgREST.Plan,174>
          \ (ds21_d2gYF :: FieldName) ->
            hpc<PostgREST.Plan,168>
            case ds20_d2gYE of {
              Nothing ->
                hpc<PostgREST.Plan,157>
                Data.Either.Left
                  (hpc<PostgREST.Plan,156> PostgREST.ApiRequest.Types.NotFound);
              Just table_a2gf2 ->
                hpc<PostgREST.Plan,167>
                hpc<PostgREST.Plan,160>
                case resolveTableField
                       (hpc<PostgREST.Plan,158> table_a2gf2)
                       (hpc<PostgREST.Plan,159> ds21_d2gYF)
                of {
                  Nothing ->
                    hpc<PostgREST.Plan,164>
                    $ (hpc<PostgREST.Plan,161>
                       \ (ds22_d2gYL :: ApiRequestError) ->
                         Data.Either.Left ds22_d2gYL)
                      (hpc<PostgREST.Plan,163>
                       PostgREST.ApiRequest.Types.ColumnNotFound
                         (hpc<PostgREST.Plan,162> ds21_d2gYF));
                  Just typedField_a2gf4 ->
                    hpc<PostgREST.Plan,166>
                    Data.Either.Right (hpc<PostgREST.Plan,165> typedField_a2gf4)
                }
            }))
      (hpc<PostgREST.Plan,177>
       S.toList (hpc<PostgREST.Plan,176> ds11_d2h02)) } in
```

So `typedColumnsOrError_a2geX` applies `mapStopOnLeft_r2g9q` on a lambda which captures `ds20_d2gYE` from a `let` statement. `ds20_d2gYE` is basically the same as your `tbl = ...`. The first time we evaluate `case ds20_d2gYE of ...` in the lambda, I believe the "thunk" for `ds20_d2gYE` on the heap [should be overwritten](https://downloads.haskell.org/~ghc/7.8.3/docs/html/users_guide/informal-semantics.html) by a "WHNF", and so not evaluated again on subsequent calls to the same lambda. (And if I'm wrong here, the proposed version shouldn't make any difference anyhow since the only difference is that the binding happens a little earlier but still outside of the lambda).

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs:341]: Nice. Yeah, just conforming to the existing structure here, agree converting back is wasted CPU cycles.

File: src/PostgREST/Query/QueryBuilder.hs
Comment (aljungberg) [reply src/PostgREST/Query/QueryBuilder.hs:97]: Yes, I think you're right.

**Comment by aljungberg** [comment]
> I have most of my suggested changes locally already. If you want to save some effort, you can check the "allow edits of maintainers" box (not sure what the name is exactly). Then I can push those to your branch.

Let's revisit that shortly. I'm trying to minimise the amount of rebasing/conflict resolution I have to do on the follow up branch.

(On that note, looking like RPC support will be easy after the rewrite. Some stuff got a little more wordy and verbose but the core will now be a "type X -> type Y" conversion engine with user defined values. This generalises to RPC parameters and results in theory. I'm not implementing that ATM but should be easy when we get there.)

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs:381]: Oh. Newbie mistake. `traverse` makes perfect sense, thanks!

(As a non-Haskell guy, it's not always obvious to me what's implicit in the behaviours of various types. Like in this case how the Functor and Applicative implementations for `Either` lead to a nice short-circuit behaviour when traversing and hitting that first `Left`.)

File: src/PostgREST/Plan.hs
Comment (aljungberg) [reply src/PostgREST/Plan.hs]: Broke out `tbl` anyhow, doesn't hurt to be more explicit that we want single eval on that expression.

File: src/PostgREST/Error.hs
Comment (aljungberg) [reply src/PostgREST/Error.hs:204]: Oh yeah, nice, quite helpful indeed! Save that work for another PR down the road?

Comment (aljungberg) [reply src/PostgREST/Error.hs]: Done and pushed.

File: src/PostgREST/Plan/Types.hs
Comment (aljungberg) [reply src/PostgREST/Plan/Types.hs]: Done.

File: src/PostgREST/Plan/Types.hs
Comment (aljungberg) [reply src/PostgREST/Plan/Types.hs]: Done including the rename to `tfName`.

**Comment by aljungberg** [comment]
I believe all feedback up to this point has been addressed.

File: src/PostgREST/Error.hs
Comment (steve-chavez) [reply src/PostgREST/Error.hs]: Sorry for the late reply here.

This is something that still worries me, I've got many users at Supabase that require hand-holding and removing this error is no good for them.

Could you try to add a hint for columns with similar names? On https://github.com/PostgREST/postgrest/pull/2575 we're using the `FuzzySet` library, that might help you get started.

File: src/PostgREST/Error.hs
Comment (steve-chavez) [inline src/PostgREST/Error.hs:202]: Making the message resemble postgresql (`column projects.xxname does not exist"`):

```suggestion
    "message" .= ("Column '" <> colName <> "' does not exist" :: Text),
```

Adding the table name would be great too.

File: src/PostgREST/Error.hs
Comment (aljungberg) [reply src/PostgREST/Error.hs]: Sorry, I just noticed this: are we talking about errors when using unknown columns with `?select`? This PR only changes the error for unknown `?columns`. The behaviour for `select` remains the same:

```
curl 'localhost:3000/todos?select=id,xdone'
# before
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"todos.done\".","message":"column todos.xdone does not exist"}
# after
{"code":"42703","details":null,"hint":"Perhaps you meant to reference the column \"todos.done\".","message":"column todos.xdone does not exist"}
```

This PR only changes the message for bad `columns`:

```
curl --request POST \
  --url 'http://localhost:3000/todos?columns=xtask' \
  --header 'Content-Type: application/json' \
  --header 'Prefer: return=representation' \
  --data '{"id":5, "task": "buy eggs", color": "#001100"}'
# before
{"code":"42703","details":null,"hint":null,"message":"column \"xtask\" of relation \"todos\" does not exist"}
# after
{"code":"PGRST118","details":null,"hint":null,"message":"Column 'xtask' does not exist"}
```

Indeed the table name was there before, but other than that pretty much the same, no fuzzy suggestions from PostgreSQL in this case.

So are we okay in that case? I certainly don't want to add to your support burden and I feel like this PR should not since the behaviour is the same before and after.

File: src/PostgREST/Error.hs
Comment (aljungberg) [reply src/PostgREST/Error.hs]: I've made the change in message style and added the table name, both (separate from our talk about fuzzy matching above). Thanks!

File: src/PostgREST/Error.hs
Comment (steve-chavez) [reply src/PostgREST/Error.hs]: Oh, I see. I was thinking that PostgreSQL suggestions for `SELECT` would apply to insertions as well.

The table name on the error message would still be nicer but not critical as for now we can only do an insert on a single table at a time.

> So are we okay in that case?

Yes, it's all good now.

Comment (aljungberg) [reply src/PostgREST/Error.hs]: Table name already added! Agreed, more of a nice to have but it's more consistent with the select error.

**Review by steve-chavez — APPROVED** [review]
LGTM! The error message now is clear as the one from PostgreSQL!

**Comment by steve-chavez** [comment]
@wolfgangwalther Do you have more input on this PR? I think it's looking good to merge.

**Comment by steve-chavez** [comment]
>  It's on my list - as well as a lot of other things ;)

Oh, no problem!

>  Feel free to merge, if you think it's good.

I hesitate because I'm a bit lost on https://github.com/PostgREST/postgrest/pull/2523 (which IIUC is the main goal of the PR ), so will let you review and merge this one.

**Comment by aljungberg** [comment]
> I hesitate because I'm a bit lost on #2523 (which IIUC is the main goal of the PR ), so will let you review and merge this one.

Anything I can do to help explain or detail it for you? If you'd like I could discuss it over a video call.

**Comment by steve-chavez** [comment]
@aljungberg Sorry for the delay! Let's move forward with this one - we'll also need it for #2594, so I've just merged it.

(I'll start reviewing #2523 and I'll let you know any doubts I have there)

**Pull request merged**
